
# Trabalho Prático 03 - Bancos de Dados I – 2025/02
Turma: CB01  
Alunos:
* Luiz Henrique Barbosa Costa - 22351234  
* Nycksandro Lima dos Santos - 22351228  
* Rômulo Fernandes Torres - 22351220
        

In [27]:
!pip install ipython-sql psycopg2-binary prettytable

### **PARTE I**

Esta primeira parte do trabalho tem dois objetivos principais. O primeiro é verificar alguns parâmetros do hardware e do software utilizado e o segundo é que os estudantes preparem e se familiarizar com o ambiente do PostgreSQL


#### **Tarefa 1 – Identificação do Sistema**

Identifique o sistema que será usado para os experimentos, incluindo informações sobre o hardware e o Sistema Operacional utilizado. Sobre o tipo de processador, quantidade de memória RAM, tamanho do disco. Devem ser também apresentadas informações sobre as caches existentes.  Sobre o Sistema Operacional, que devem ser Linux, incluir informações sobre qual a distribuição usada, versão do sistema, versão do Kernel, etc.

**O que entregar:** As informações pedidas devem ser apresentadas no jupyter notebook

#### Resposta:

No terminal do linux, executar cada um desses comandos abaixos:

#### Informações do SO:

> **Distribuição e Versão:** cat /etc/os-release  
> **Versão do Kernel:** uname -r  
> **Arquitetura:** uname -m

#### Informações do Hardware

> **CPU:** cat /proc/cpuinfo | grep "model name" | head -1  
> **Caches da CPU:** lscpu | grep "cache"  
> **Mémoria RAM:** free -h  
> **Nome, tamanho e modelo dos discos:** lsblk -d -o NAME,SIZE,MODEL

Após executar no PC:
> **Distribuição e Versão:**  
> **Versão do Kernel:**  
> **Arquitetura:**
> **CPU**:  
> **Caches da CPU:**  
> **Mémoria RAM:**  
> **Tamanho do disco:**  
> **Modelo do disco:**  


#### **Tarefa 2 - Verificação de parâmetros de armazenamento**

a) Verifique no disco que será usado para os experimentos no laboratório os seguintes parâmetros: Nr. de superfícies, cilindros, setores por trilha, velocidade de rotação, latência rotacional; tempos de seek médio, máximo e mínimo; tempo para a próxima trilha; e taxa de transferência.

b) Utilizando o comando “stat” do Linux, verifique os parâmetros dos parâmetros de S.O. que serão utilizados para o disco.

c) Verifique o tamanho de bloco utilizado e mostre como alterar o tamanho dos blocos

**O que entregar:** Os resultados de cada verificação devem ser preenchidas no jupyter notebook

#### Resposta:

a)

Comandos do linux para encontrar alguns:

> **Cilindros, Supefícies e setores por trilha (mapeamento lógico):** sudo hdparm -g /dev/sda  
> obs: Esse comando vai retornar no formato **geometry  = X/Y/Z, sectors = A, start = B**, onde X é cilindros, Y é cabeças/superfícies e Z é setores por trilha)  
> **Velocidade de rotação:** sudo hdparm -I /dev/sda | grep "Rotation Rate"  
> **Latência rotacional média**: (30/Velocidade da rotação) * 1000  
> **Taxa de transferência (leitura sequencial):** sudo hdparm -t /dev/sda  


> **Número de superfícies:**  
> **Cilindros:**  
> **Setores por trilha:**  
> **Velocidade de rotação:**  
> **Latência rotacional:**  
> **Tempos de seek médio:**  
> **Tempos de seek máximo:**  
> **Tempos de seek mínimo:**  
> **Tempo para a próxima trilha:**  
> **Taxa de transferência:**  

b)

Comandos do linux para encontrar alguns:

> **Exibe os parâmetros do SO que serão utilizados no disco:** stat -f /home/  
    >> Vai exibir no seguinte formato:
        >>> Arquivo: "/home/"  
        ID: 36dedc6dd8a83f11 Comprimento nome: 255     Tipo: ext2/ext3  
        **Tamanho de bloco: 4096       Tamanho fundamental de bloco: 4096**  
        **Blocos: Total: 22948018   Livres: 9317935    Disponíveis: 8140758**    
        Inodes: Total: 5865472    Livres: 4697212  


Os parâmetros são: Tipo do Sistema de Arquivos, Tamanho de blocos de I/O, Tamanho Fundamental de Bloco, Blocos Totais, Blocos Livres, Blocos Disponíveis (a primeira linha inteira)

c) O tamanho do bloco é 4096 Bytes

In [ ]:
!sudo hdparm -g /dev/sda

#### **Tarefa 3 – Geração de um BD para testes**

Essa tarefa consiste na criação de banco de dados e povoamento destas tabelas com dados sintéticos.  Para a definição do esquema das tabelas e os dados a serem carregados usaremos a especificação e os utilitários fornecidos pelos Benchmark TPC-H (http://www.tpc.org/tpch/).

Para geração do BD no PostgreSQL, siga as instruções disponíveis em:

https://github.com/foliveirafilho/tpch-pgsql

**O que entregar:** Devem ser apresentadas as saídas da execução dos scripts de geração.

Ordem de execuções dos comandos:  
**1) Passo:** Baixar o zip do repositório https://github.com/foliveirafilho/tpch-pgsql  
**2) Passo:** Descompactar o repositório  
**3) Passo:** Executar esses comandos abaixo no terminal do Linux
> * !wget -q https://github.com/electrum/tpch-dbgen/archive/32f1c1b92d1664dba542e927d23d86ffa57aa253.zip -O tpch-dbgen.zip  
> * !unzip -q tpch-dbgen.zip && mv tpch-dbgen-32f1c1b92d1664dba542e927d23d86ffa57aa253 tpch-dbgen && rm tpch-dbgen.zip  

**4) Passo:** Colocar essa pasta que acabou de ser extraida acima dentro da pasta do Passo 2, ficando assim: **tpch-pgsql-master/tpch-dbgen**  
**5) Passo:** Dar cd nessa pasta **tpch-pgsql-master/** e executar no terminal
> * !pip3 install -r requirements.txt  

**6) Passo:** Dar cd .. e executar no terminal
> * chmod -R u+rwX tpch-dbgen
> * rm -rf ./data  
> * rm -f ./data/load/*.tbl  

**7) Passo:** Executar essa série de comandos no terminal do linux
> !sudo apt-get install -y postgresql postgresql-contrib  (Caso não tenha o postgre)

> !sudo service postgresql start  
> !sudo -u postgres createuser tpch  
> !sudo -u postgres createdb tpchdb  
> !sudo -u postgres psql -c "ALTER USER tpch WITH ENCRYPTED PASSWORD 'test123';"  
> !sudo -u postgres psql -c "GRANT ALL PRIVILEGES ON DATABASE tpchdb TO tpch;"  
    
**8) Passo:** Executar o script que implementa todas as fases do benchmark
> * python3 tpch_pgsql.py prepare -g ./tpch-dbgen -s 1.0
> * psql -h localhost -p 5432 -U tpch -d tpchdb -W  

**9) Passo:** Caso dê erro no passo 8, será necessário mudar as permissões da pasta tpch-pgsql-master para permitir leitura e escrita, efetuando isso para todos os arquivos dela também. Após isso provavelmente irá funcionar.

In [34]:
!git clone https://github.com/foliveirafilho/tpch-pgsql.git

fatal: destination path 'tpch-pgsql' already exists and is not an empty directory.


In [33]:
!wget -q https://github.com/electrum/tpch-dbgen/archive/32f1c1b92d1664dba542e927d23d86ffa57aa253.zip -O tpch-dbgen.zip
!unzip -q tpch-dbgen.zip && mv tpch-dbgen-32f1c1b92d1664dba542e927d23d86ffa57aa253 tpch-dbgen && rm tpch-dbgen.zip

In [40]:
!mv tpch-dbgen tpch-pgsql/

mv: cannot stat 'tpch-dbgen': No such file or directory


In [41]:
%cd tpch-pgsql/tpch-dbgen
!pip3 install -r requirements.txt

[Errno 2] No such file or directory: 'tpch-pgsql/tpch-dbgen'
/content/tpch-pgsql-master
ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'requirements.txt'


In [32]:
!sudo service postgresql start

 * Starting PostgreSQL 14 database server
   ...done.


In [28]:
#Nessa célula a gente estabelece a conexão com o BD

import pandas as pd
from sqlalchemy import create_engine, text

# 1. Carrega a extensão SQL
%load_ext sql

# 2. String de conexão
conexao_string = "postgresql://tpch:test123@localhost:5432/tpchdb"

# 3. Conecta ao banco (o comando '%sql' executa a conexão)
%sql $conexao_string

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/sqlalchemy/engine/base.py", line 143, in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sqlalchemy/engine/base.py", line 3301, in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sqlalchemy/pool/base.py", line 447, in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sqlalchemy/pool/base.py", line 1264, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sqlalchemy/pool/base.py", line 711, in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
  F

In [31]:
!pgrep -l postgres

In [ ]:
#Criando uma engine para poder executar as query e imprimir o resultado nas células
engine = create_engine(conexao_string)

def executa_consulta(sql_query): # Função que criamos apenas para executar as query e fazer o resultado aparecer nas células
    try:
        with engine.connect() as connection:
            result = connection.execute(text(sql_query)).fetchall()
        print("--------------------------------------------------------------------------------------------------------------------------------------------------------")
        for row in result:
            print(row[0])
        print("--------------------------------------------------------------------------------------------------------------------------------------------------------")

    except Exception as e:
        print(f"ERRO: Não foi possível executar a consulta. {e}")

#### **Tarefa 4 – Execução de Consultas**

A segunda tarefa deste trabalho consiste em executar e analisar um conjunto de consultas analíticas do benchmark TPC-H sobre o banco de dados gerado na Tarefa 1, observando o comportamento do PostgreSQL em diferentes tipos de operações (filtros, junções, agregações e ordenações). O objetivo é compreender como o otimizador escolhe planos de execução e medir o tempo de resposta de consultas típicas de um sistema de apoio à decisão.

O TPC-H define 22 consultas padronizadas (Q1 a Q22), mas para este trabalho será utilizado um subconjunto representativo composto por 8 consultas que abrangem diferentes padrões de acesso.

Consultas a executar: Q1, Q3, Q5, Q6, Q7, Q9, Q10 e Q12. As consultas originais podem ser obtidos no repositório oficial do benchmark. Essas consultas devem ser executadas no mesmo banco TPC-H criado na Tarefa 1, utilizando o PostgreSQL.

Para cada consulta:
* Execute o comando SQL completo no PostgreSQL.
* Mostre as 10 primeiras linhas do resultado (ou um trecho representativo).
* Execute o comando EXPLAIN ANALYZE antes da consulta para exibir o plano de execução e o tempo real medido O comando EXPLAIN ANALYZE mostra as etapas internas do plano escolhido pelo otimizador, o número de linhas processadas e o tempo de cada operação.
* Registre o tempo total de execução e comente brevemente o tipo de acesso observado (por exemplo, varredura sequencial, uso de índice, junções em hash, etc.).

**O que entregar:** Devem ser apresentados os resultados acima

### **PARTE II**

O objetivo desta parte do trabalho é analisar o comportamento dos índices das tabelas do SGBD através do exame e análise das tabelas de estatísticas para consultas SQL sobre uma tabela criada com dados aleatórios.


#### **Tarefa 5 – Preparação da Tabela Aleatória**
Criar uma tabela com uma chave simples e alguns dados de exemplo. Cada valor de chave é um número incremental e está associado a com valores que variam de 0 até 10:

DROP TABLE IF EXISTS t;
CREATE TABLE t (k serial PRIMARY KEY, v integer);

INSERT INTO t(v)
SELECT trunc(random() * 10)  **FROM generate_series(1,100000);**

**O que entregar:** Imprimir os valores das 10 primeiras tuplas da tabela, ordenando por k.

#### **Tarefa 6 – Páginas criadas**


Verifique quantas páginas com blocos foram criadas para a tabela da Tarefa 5.

Commando: **SELECT relname, relpages, reltuples FROM pg_class WHERE relname='t';**

**O que entregar:** Imprimir o resultado do comando SQL.


#### **Tarefa 7 – Blocos**

Verifique quantos blocos foram efetivamente usados numa consulta

**Comando:  
SELECT pg_sleep(1);  
\pset x on  
SELECT * FROM pg_stats WHERE relname='t';  
SELECT pg_stat_reset();  
\pset x off**

Observação: Em algumas versões do PostgreSQL, o atributo é chamado de **tablename** em vez de **relname**.

**O que entregar:** Imprimir o resultado do comando SQL.

#### **Tarefa 8 – Índice**

Crie um índice para o atributo ‘v’ e realize consultas e criação de índice

Qual o tempo gasto para realizar uma consulta para um valor (tendo a tabela 100000 tuplas)?
Qual o tempo gasto para recriar um índice para o atributo ‘v’?

Remova a tabela ‘t’ e crie novamente com 1.0000.000 de tuplas

Qual o tempo gasto para realizar uma consulta para um valor específico?
Qual o tempo gasto para recriar um índice para o atributo ‘v’?

**O que entregar:** Relatório com o resultado das perguntas

#### **Tarefa 9 - Fill factor**

Quando se cria um novo índice, nem toda entrada no bloco do índice é usada. Um espaço livre é deixado, conforme o parâmetro **fillfactor**.

Crie novos índices usando fillfactor=60,80,90 e 100. Analise o desempenho de suas consultas usando as mesmas condições da Tarefa 12

ALTER TABLE foo SET ( fillfactor = 50);
VACUUM FULL foo;

**O que entregar:** Relatório com o resultado das perguntas

#### **Tarefa 10 - Utilize índices com ordem DESC**

Repita os testes das Tarefas 8 e 9 usando índices descendentes. Avalie e registre o resultado

Comando: **CREATE INDEX i ON t(v DESC NULLS FIRST);**

**O que entregar:** Relatório com o resultado da avaliação e uma análise dos resultados.

### **PARTE III.**

O objetivo desta parte do trabalho é estudar o comportamento dos otimizadores de consulta dos SGBDs através do exame e análise dos planos de execução para consultas SQL sobre tabelas que serão fornecidos. Será bastante utilizado o comando EXPLAIN ANALYZE, que permite visualizar todas as etapas envolvidas no processamento de uma consulta. Usaremos para isso a tabela “movies”.

#### **Tarefa 11 - Preparação e Verificação do Ambiente**
movies: https://drive.google.com/file/d/1W6wovSsVu4B0OIo_tsSBBHi8WRKQqnat/view?usp=drive_link  

> **a)** Execute o script movie.sql em movies para criar as tabelas e índices e carregar os dados necessários às próximas atividades

>**b)** Verifique no catálogo do banco de dados os seguintes metadados sobre os índices associados às tabelas e apresente-os no relatório: Nome do índice, nome da tabela, altura, número máximo de chaves por bloco, número médio de chaves por bloco, número de blocos folha, número de médio de blocos folha por chave, número médio de blocos de dados por chave, número de linhas e número de chaves distintas.

**O que entregar:** Relatório com os resultados da verificação

##### **Como fizemos a prepação:**
**a)** Executar o comando abaixo com o movie.sql na pasta raiz.
> Apagar o ultimo BD criado da parte 1 e reutilizar ele  
> psql -h localhost -p 5432 -U tpch -d tpchdb -W -f movie.sql    

**b)** Para ver os metadados dos índices associados a tabela é preciso executar o comando abaixo no postgre
1) Para encontrar as informações sobre os títulos dos índices associados a tabela
    > CREATE EXTENSION IF NOT EXISTS pgstattuple;  
    > **\d movie** (para exibir os títulos dos índices)    

2) Para encontrar a informações da quantidade de linhas e número de chaves distintas:
    > SELECT COUNT(ID) FROM movie;

3) Para cada um dos índices, se obtém essas informações através dos seguintes comandos:
> SELECT * FROM pgstatindex('movie_key');
> SELECT * FROM pgstatindex('movie_title');
> SELECT * FROM pgstatindex('movie_votes');

#### **Informações gerais**:

> * **Nome dos índices:** movie_key, movie_title, movie_votes  
> * **Nome da tabela:**: movie  
> * **Número de linhas:** 1844  
> * **Número de chaves distintas:** 1844   

##### **movie_key**

> * **Altura da árvore:** 1  
> * **Número máximo de chaves:**  
> * **Número médio de chaves por bloco:**  
> * **Número de blocos folha:** 5  
> * **Número médio de blocos folha por chave:**
> * **Número médio de blocos de dados por chave:**  


##### **movie_title**

> * **Altura da árvore:**  
> * **Número máximo de chaves:**  
> * **Número médio de chaves por bloco:**  
> * **Número de blocos folha:**  
> * **Número médio de blocos folha por chave:**  
> * **Número médio de blocos de dados por chave:**

##### **movie_votes**

> * **Altura da árvore:**  
> * **Número máximo de chaves:**  
> * **Número médio de chaves por bloco:**  
> * **Número de blocos folha:**  
> * **Número médio de blocos folha por chave:**  
> * **Número médio de blocos de dados por chave:**

#### **Tarefa 12 - Consultas por intervalo e índices secundários**
>**a)** Escreva uma consulta em SQL sobre o atributo VOTES da tabela MOVIE que recupera um número pequeno de tuplas (<10 tuplas); Execute o comando EXPLAIN ANALYZE sobre esta consulta e apresente os resultados.  

>**b)** Escreva uma consulta em SQL sobre o atributo VOTES da tabela MOVIE que recupera um número grande de tuplas (>80% das tuplas). Execute o comando EXPLAIN ANALYZE sobre esta consulta e apresente os resultados.

>**c)** Explique porque o índice sobre VOTES não é sempre usado nas consultas sobre este atributo.  

**O que entregar:** Relatório com as respostas das questões.

In [ ]:
#a)
consulta = """
EXPLAIN ANALYZE
SELECT * FROM movie WHERE Votes > 1483 and Votes < 1500;
"""

executa_consulta(consulta)


--------------------------------------------------------------------------------------------------------------------------------------------------------
Index Scan using movie_votes on movie  (cost=0.28..8.42 rows=7 width=30) (actual time=0.029..0.047 rows=6 loops=1)
  Index Cond: ((votes > 1483) AND (votes < 1500))
Planning Time: 0.176 ms
Execution Time: 0.099 ms
--------------------------------------------------------------------------------------------------------------------------------------------------------


#### **Resposta da letra a)**
Retorna 6 linhas e utiliza o índice sobre VOTES, tendo como resultado:
> Index Scan using movie_votes on movie  (cost=0.28..8.42 rows=7 width=30) (actual time=0.044..0.061 rows=6 loops=1)  
> Index Cond: ((votes > 1483) AND (votes < 1500))  
> Planning Time: 0.201 ms  
> Execution Time: 0.162 ms  


In [ ]:
#b) Executando o comando a seguir

consulta = """
EXPLAIN ANALYZE
SELECT *  FROM movie WHERE Votes > 865;
"""

executa_consulta(consulta)

--------------------------------------------------------------------------------------------------------------------------------------------------------
Seq Scan on movie  (cost=0.00..38.05 rows=1672 width=30) (actual time=0.015..3.435 rows=1673 loops=1)
  Filter: (votes > 865)
  Rows Removed by Filter: 171
Planning Time: 0.105 ms
Execution Time: 6.606 ms
--------------------------------------------------------------------------------------------------------------------------------------------------------


#### **Resposta da letra b)**

Como há 1844 linhas nessa tabela, para retornar 90% das tuplas, preciso de pelo menos 1670 tuplas. A consulta acima retornou 1673 tuplas, não utilizando o índice sobre VOTES e tem como resultado:    

> Seq Scan on movie  (cost=0.00..38.05 rows=1672 width=30) (actual time=0.024..3.700 rows=1673 loops=1)  
> Filter: (votes > 865)  
> Rows Removed by Filter: 171    
> Planning Time: 0.158 ms  
> Execution Time: 7.188 ms  

#### **Resposta da letra c)**

 Como na letra a), a consulta tem uma alta seletividade e retorna um número muito pequeno de tuplas em relação ao total, o otimizador calcula previamente e toma a decisão de utilizar o índice pois o custo de navegar no índice e buscar as pouquissimas linhas seria bem menor do que varrer a tabela inteira, dessa forma, o otimizador escolhe o Index Scan. Já a letra b), é uma consulta com uma baixa seletividade, retornando um número muito grande de tuplas (quase a tabela inteira), logo o otimizador, calcula previamente e toma a decisão de não utilizar o índice, varrendo a tabela inteira de forma sequencial, evitando assim muitos acessos aleatórios, e por conta disso, utiliza o Seq Scan.

#### **Tarefa 13 - Comparações de operadores de agregação.**

Considere as seguintes consultas em SQL, sobre o atributo VOTES, as quais são equivalentes:

* SELECT title FROM movie WHERE votes >= (SELECT MAX(votes) FROM movie);
* SELECT title FROM movie WHERE votes >= ALL (SELECT votes FROM movie) ;

>**a)** Apresente o resultado do comando EXPLAIN ANALYZE sobre as duas consultas acima  
>**b)** Existe alguma diferença entre os planos de consultas? Qual das duas é mais eficiente? Explique?

**O que entregar:** Relatório com as respostas das questões.


In [ ]:
#a)

#1) Consulta:

consulta1 = """
EXPLAIN ANALYZE
SELECT title FROM movie WHERE votes >= (SELECT MAX(votes) FROM movie);
"""

executa_consulta(consulta1)


--------------------------------------------------------------------------------------------------------------------------------------------------------
Index Scan using movie_votes on movie  (cost=0.61..35.37 rows=615 width=16) (actual time=0.105..0.121 rows=1 loops=1)
  Index Cond: (votes >= $1)
  InitPlan 2 (returns $1)
    ->  Result  (cost=0.32..0.33 rows=1 width=4) (actual time=0.078..0.089 rows=1 loops=1)
          InitPlan 1 (returns $0)
            ->  Limit  (cost=0.28..0.32 rows=1 width=4) (actual time=0.069..0.076 rows=1 loops=1)
                  ->  Index Only Scan Backward using movie_votes on movie movie_1  (cost=0.28..76.55 rows=1844 width=4) (actual time=0.064..0.066 rows=1 loops=1)
                        Index Cond: (votes IS NOT NULL)
                        Heap Fetches: 0
Planning Time: 0.327 ms
Execution Time: 0.175 ms
-------------------------------------------------------------------------------------------------------------------------------------------------

In [ ]:
#a)

#2) Consulta:

consulta2 = """
EXPLAIN ANALYZE
SELECT title FROM movie WHERE votes >= ALL (SELECT votes FROM movie);
"""

executa_consulta(consulta2)

--------------------------------------------------------------------------------------------------------------------------------------------------------
Seq Scan on movie  (cost=0.00..43620.99 rows=922 width=16) (actual time=14.106..21.005 rows=1 loops=1)
  Filter: (SubPlan 1)
  Rows Removed by Filter: 1843
  SubPlan 1
    ->  Materialize  (cost=0.00..42.66 rows=1844 width=4) (actual time=0.002..0.008 rows=2 loops=1844)
          ->  Seq Scan on movie movie_1  (cost=0.00..33.44 rows=1844 width=4) (actual time=0.006..3.639 rows=1844 loops=1)
Planning Time: 0.112 ms
Execution Time: 21.064 ms
--------------------------------------------------------------------------------------------------------------------------------------------------------


#### **Resposta da letra a)**

##### Para a primeira consulta: Retorna 1 linha que satisfaz a condição

Index Scan using movie_votes on movie  (cost=0.61..35.37 rows=615 width=16) (actual time=0.174..0.190 rows=1 loops=1)  
>Index Cond: (votes >= $1)  
>InitPlan 2 (returns $1)  
    >> -> Result  (cost=0.32..0.33 rows=1 width=4) (actual time=0.146..0.157 rows=1 loops=1)  
        >>>InitPlan 1 (returns $0)
            >>>> -> Limit  (cost=0.28..0.32 rows=1 width=4) (actual time=0.135..0.143 rows=1 loops=1)
                >>>>> ->  Index Only Scan Backward using movie_votes on movie movie_1  (cost=0.28..76.55 rows=1844 width=4) (actual time=0.131..0.133 rows=1 loops=1)   
                    >>>>>> Index Cond: (votes IS NOT NULL)  
                    >>>>>> Heap Fetches: 0  

 Planning Time: 73.728 ms  
 Execution Time: 0.352 ms

##### Para a segunda consulta: Retorna 1 linha que satisfaz a condição

Seq Scan on movie  (cost=0.00..43620.99 rows=922 width=16) (actual time=15.678..23.853 rows=1 loops=1)  
   >Filter: (SubPlan 1)  
   >Rows Removed by Filter: 1843  
   >SubPlan 1  
    >> ->  Materialize  (cost=0.00..42.66 rows=1844 width=4) (actual time=0.002..0.008 rows=2 loops=1844)  
        >>>   ->  Seq Scan on movie movie_1  (cost=0.00..33.44 rows=1844 width=4) (actual time=0.006..4.026 rows=1844 loops=1)  

 Planning Time: 0.142 ms  
 Execution Time: 24.045 ms  

#### **Resposta da letra b)**

Sim, existe. Ambas utilizam estratégias bem diferentes:
* Primeira Consulta: Utiliza duas subconsultas em índices. A primeira executa Index Only Scan Backward, que acessa o índice e percorre no sentido inverso até encontrar o valor máximo de votos, guardando na váriavel $0 e disponibilizando para o InitPlan2 como a váriavel $1. Após isso, é executado um Index Scan para encontrar os votos maiores que o encontrado no primeiro Index Scan.
* Segunda Consulta: Primeiramente, é executado um Seq Scan como filtro, onde é varrido a tabela inteira e o resultado é salvo temporariamente como movie_1 atráves do Materialize, apoś isso é executado outro Seq Scan para varrer a tabela movie inteira e para cada linha lida, é utilizada o movie_1 para a aplicação do filtro.

A primeira consulta é muito mais eficiente do que a segunda (cerca de 68 vezes mais rápido), tal afirmação pode ser comprovada observando as estatísticas cedidas na consulta. Tanto o custo estimado pelo otimizador, quanto o tempo de execução são bem menores na primeira consulta, além disso, a estratégia adotada na consulta 1 irá acessar menos a mémoria secundária que na consulta 2. Porém, é válido ressaltar, o custo associado a verificação de cada plano, sendo o custo de verificação da segunda consulta bem menor que o da primeira (cerca de 519 vezes menor).

#### **Tarefa 14 - Consultas com Junção e Seleção**

Considere as duas consultas equivalentes em SQL a seguir, as quais retornam os filmes com mais votos que “Star Wars”

* SELECT title FROM movie WHERE votes > (SELECT votes FROM movie WHERE title = 'Star Wars');
* SELECT m1.title FROM movie m1, movie m2 WHERE m1.votes > m2.votes AND m2.title = 'Star Wars';

>**a)** Apresente o resultado do comando EXPLAIN ANALYZE sobre as duas consultas acima  
>**b)** Existe alguma diferença entre os planos de consultas? Qual das duas é mais eficiente? Explique?

**O que entregar:** Relatório com as respostas das questões.


In [ ]:
# a)

# 1) Consulta:

consulta1 = """
EXPLAIN ANALYZE
SELECT title FROM movie WHERE votes > (SELECT votes FROM movie WHERE title = 'Star Wars');
"""

executa_consulta(consulta1)

--------------------------------------------------------------------------------------------------------------------------------------------------------
Index Scan using movie_votes on movie  (cost=8.57..43.34 rows=615 width=16) (actual time=0.055..0.061 rows=0 loops=1)
  Index Cond: (votes > $0)
  InitPlan 1 (returns $0)
    ->  Index Scan using movie_title on movie movie_1  (cost=0.28..8.29 rows=1 width=4) (actual time=0.034..0.039 rows=1 loops=1)
          Index Cond: ((title)::text = 'Star Wars'::text)
Planning Time: 0.229 ms
Execution Time: 0.094 ms
--------------------------------------------------------------------------------------------------------------------------------------------------------


In [ ]:
# a)

# 2) Consulta:

consulta2 = """
EXPLAIN ANALYZE
SELECT m1.title FROM movie m1, movie m2 WHERE m1.votes > m2.votes AND m2.title = 'Star Wars';
"""

executa_consulta(consulta2)

--------------------------------------------------------------------------------------------------------------------------------------------------------
Nested Loop  (cost=0.56..49.49 rows=615 width=16) (actual time=0.038..0.047 rows=0 loops=1)
  ->  Index Scan using movie_title on movie m2  (cost=0.28..8.29 rows=1 width=4) (actual time=0.017..0.021 rows=1 loops=1)
        Index Cond: ((title)::text = 'Star Wars'::text)
  ->  Index Scan using movie_votes on movie m1  (cost=0.28..35.04 rows=615 width=20) (actual time=0.007..0.009 rows=0 loops=1)
        Index Cond: (votes > m2.votes)
Planning Time: 0.373 ms
Execution Time: 0.081 ms
--------------------------------------------------------------------------------------------------------------------------------------------------------


#### **Resposta da letra a)**

#### Para a primeira consulta: Retorna 0 linhas

Index Scan using movie_votes on movie  (cost=8.57..43.34 rows=615 width=16) (actual time=42.950..42.957 rows=0 loops=1)  
   > Index Cond: (votes > $0)  
   > InitPlan 1 (returns $0)  
    >> ->  Index Scan using movie_title on movie movie_1  (cost=0.28..8.29 rows=1 width=4) (actual time=42.920..42.928 rows=1 loops=1)  
        >>>   Index Cond: ((title)::text = 'Star Wars'::text)  

Planning Time: 0.172 ms  
Execution Time: 42.988 ms  

#### Para a segunda consulta: Retorna 0 linhas

Nested Loop  (cost=0.56..49.49 rows=615 width=16) (actual time=0.038..0.047 rows=0 loops=1)  
   > ->  Index Scan using movie_title on movie m2  (cost=0.28..8.29 rows=1 width=4) (actual time=0.018..0.022 rows=1 loops=1)  
        >> Index Cond: ((title)::text = 'Star Wars'::text)  

   > ->  Index Scan using movie_votes on movie m1  (cost=0.28..35.04 rows=615 width=20) (actual time=0.007..0.009 rows=0 loops=1)  
        >> Index Cond: (votes > m2.votes)  

 Planning Time: 81.866 ms
 Execution Time: 0.079 ms


#### **Resposta da letra b)**
Sim, existe. Embora ambas as consultas utilize índices sobre os votos, as suas estratégias são diferentes.
* Primeira Consulta: Executar Index Scan para encontrar os votos do filme "Star Wars", guardando o valor dos votos na variável $0. Somente após essa subconsulta ser executada e encerrada, é realizada outro Index Scan para encontrar os votos maiores que $0
* Segunda Consulta: Executa um Index Scan para encontrar os votos do filme com o nome de "Star Wars", e também Executando outro Index Scan para encontrar pelos votos maiores que o encontrado no Index Scan anterior.

Em questão de desempenho, a segunda consulta apresentou um tempo de execução muito menor (cerca de 544 vezes menor), muito provavelmente devido ao fato do uso do Nested Loop e também por conta do PostgreSQL consegue executar os dois Index Scan de forma mais eficiente do que na primeira consulta, passando os dados em forma de pipeline em em vez de materializar, evitando assim gastos desnecessários. Porém, é válido ressaltar que o tempo de planejamento necessário foi 476 vezes maior.

#### **Tarefa 15 - Casamento de Strings e Índices**

Considere as seguintes consultas SQL sobre o atributo  TITLE usando o operador LIKE.

* SELECT title FROM movie WHERE title LIKE 'I%';
* SELECT title FROM movie WHERE substr(title, 1, 1) = 'I';
* SELECT title FROM movie WHERE title LIKE '%A';

>**a)** Apresente o resultado do comando explain sobre as três consultas acima  
>**b)** Qual das três apresenta o menor custo? Porque?  
>**c)** O índice sobre TITLE foi usado para todas elas? Justifique.

**O que entregar:** Relatório com as respostas das questões.


In [ ]:
#a)

#1) Consulta:

consulta1 = """
EXPLAIN ANALYZE
SELECT title FROM movie WHERE title LIKE 'I%';
"""

executa_consulta(consulta1)

--------------------------------------------------------------------------------------------------------------------------------------------------------
Seq Scan on movie  (cost=0.00..38.05 rows=18 width=16) (actual time=0.020..0.579 rows=25 loops=1)
  Filter: ((title)::text ~~ 'I%'::text)
  Rows Removed by Filter: 1819
Planning Time: 0.185 ms
Execution Time: 0.657 ms
--------------------------------------------------------------------------------------------------------------------------------------------------------


In [ ]:
#a)

#2) Consulta:

consulta2 = """
EXPLAIN ANALYZE
SELECT title FROM movie WHERE substr(title, 1, 1) = 'I';
"""

executa_consulta(consulta2)

--------------------------------------------------------------------------------------------------------------------------------------------------------
Seq Scan on movie  (cost=0.00..42.66 rows=9 width=16) (actual time=0.023..0.901 rows=25 loops=1)
  Filter: (substr((title)::text, 1, 1) = 'I'::text)
  Rows Removed by Filter: 1819
Planning Time: 0.088 ms
Execution Time: 0.976 ms
--------------------------------------------------------------------------------------------------------------------------------------------------------


In [ ]:
#a)

#3) Consulta;

consulta3 = """
EXPLAIN ANALYZE
SELECT title FROM movie WHERE title LIKE '%A';
"""
executa_consulta(consulta3)

--------------------------------------------------------------------------------------------------------------------------------------------------------
Seq Scan on movie  (cost=0.00..38.05 rows=18 width=16) (actual time=0.031..0.583 rows=30 loops=1)
  Filter: ((title)::text ~~ '%A'::text)
  Rows Removed by Filter: 1814
Planning Time: 0.135 ms
Execution Time: 0.668 ms
--------------------------------------------------------------------------------------------------------------------------------------------------------


#### **Resposta da letra a)**

#### Para a primeira consulta: Retorna 25 linhas que satisfazem a condição

Seq Scan on movie  (cost=0.00..38.05 rows=18 width=16) (actual time=0.030..0.430 rows=25 loops=1)  
   > Filter: ((title)::text ~~ 'I%'::text)  
   > Rows Removed by Filter: 1819  

Planning Time: 0.153 ms  
Execution Time: 0.509 ms  


#### Para a segunda consulta: Retorna 25 linhas que satisfazem a condição
Seq Scan on movie  (cost=0.00..42.66 rows=9 width=16) (actual time=0.038..0.915 rows=25 loops=1)  
   > Filter: (substr((title)::text, 1, 1) = 'I'::text)  
   > Rows Removed by Filter: 1819  

Planning Time: 0.147 ms  
Execution Time: 0.995 ms  

#### Para a terceira consulta: Retorna 30 linhas que satisfazem a condição

Seq Scan on movie  (cost=0.00..38.05 rows=18 width=16) (actual time=0.024..0.527 rows=30 loops=1)  
   > Filter: ((title)::text ~~ '%A'::text)  
   > Rows Removed by Filter: 1814  

Planning Time: 0.105 ms  
Execution Time: 0.634 ms  

#### **Resposta da letra b)**

Dentre as 3 consultas, as consultas 1 e 3 apresentam o mesmo custo estimado de 38.05, e são menores que a consulta 2 que apresentou 42.66 de custo estimado. Isso ocorre pois a consulta 2 utiliza "SUBSTR(x,y,z)", uma função completa que extrai z caracteres de uma string x, a partir da posição y, logo ela gerará mais overhead que as outras consultas que utilizam "LIKE". Dentre as consultas 1 e 3, a que apresentou o menor custo é a consulta 1, que possui o menor tempo de execução.

#### **Resposta da letra c)**

Não, o índice não foi utilizada em nenhuma das 3 consultas.
* Primeira Consulta: Embora fosse totalmente possível utilizar índices, uma vez que a condição era "I%", correpondente a letra "I" seguida de qualquer outra, e os índices estão organizado de ordem alfabética, o otimizador provavelmente decidiu utilizar Seq Scan devido ao tamanho da tabela ser pequeno.
* Segunda Consulta: Como é uma função que vai receber como primeiro parâmetro (string a ser extraída) o título da linha, logo não é possível utilizar índices para isso, dessa forma, é utilizado Seq Scan.
* Terceira Consulta: Não foi utilizada pois a condição "%A" corresponde a letra coringa seguida da letra "A", e como o índice é organizado de forma alfabética, isso torna impossível utilizar índices nesse caso.

#### **Tarefa 16 - Verificação da hipótese de distribuição uniforme na estimativa de seletividade**

Considere as seguintes  consultas sobre o atributo TITLE da tabela MOVIE

* SELECT title FROM movie WHERE votes < 1000;
* SELECT title FROM movie WHERE votes > 40000;

>**a)** Apresente o resultado do comando explain sobre as duas consultas acima. Explique o resultado.  
>**b)** Compare o número de tuplas selecionadas por cada consulta. Qual das duas tem a menor seletividade?

**O que entregar:** Relatório com as respostas das questões.

In [ ]:
# a)

#1) Consulta:

consulta1 = """
EXPLAIN ANALYSE
SELECT title FROM movie WHERE votes < 1000;
"""

executa_consulta(consulta1)

--------------------------------------------------------------------------------------------------------------------------------------------------------
Index Scan using movie_votes on movie  (cost=0.28..20.02 rows=328 width=16) (actual time=0.033..0.773 rows=326 loops=1)
  Index Cond: (votes < 1000)
Planning Time: 0.096 ms
Execution Time: 1.394 ms
--------------------------------------------------------------------------------------------------------------------------------------------------------


In [ ]:
#a)

#2) Consulta:

consulta2 = """
EXPLAIN ANALYSE
SELECT title FROM movie WHERE votes > 40000;
"""

executa_consulta(consulta2)

--------------------------------------------------------------------------------------------------------------------------------------------------------
Index Scan using movie_votes on movie  (cost=0.28..8.42 rows=8 width=16) (actual time=0.013..0.025 rows=4 loops=1)
  Index Cond: (votes > 40000)
Planning Time: 0.128 ms
Execution Time: 0.057 ms
--------------------------------------------------------------------------------------------------------------------------------------------------------


#### **Resposta da letra a)**

#### Primeira Consulta: Retorna 326 linhas que satisfazem a condição
Utiliza Index Scan para realizar a consulta, tendo o custo estimado de 20.02, realizando a execução da consulta em 1.365 ms e o tempo de planejamento da consulta em 0.093 ms


Index Scan using movie_votes on movie  (cost=0.28..20.02 rows=328 width=16) (actual time=0.025..0.750 rows=326 loops=1)  
   > Index Cond: (votes < 1000)  

Planning Time: 0.093 ms  
Execution Time: 1.365 ms  

#### Segunda Consulta: Retorna apenas 4 linhas que satisfazem essa condição
Também utiliza Index Scan para realizar a consulta, tendo o custo estimado de apenas 8.42, realizando a execução da consulta em 0.091 ms e o planejamento da consulta em 0.115 ms

Index Scan using movie_votes on movie  (cost=0.28..8.42 rows=8 width=16) (actual time=0.013..0.025 rows=4 loops=1)  
  > Index Cond: (votes > 40000)  

Planning Time: 0.115 ms  
Execution Time: 0.091 ms  


#### **Resposta da letra b**

Enquanto a primeira consulta retorna 326 linhas, a segunda consulta retorna somente 4 linhas. Com isso, é possível afirmar que a primeira consulta possui uma menor seletividade por retornar um número maior de linhas.

### **PARTE IV**

O objetivo desta parte do trabalho é experimentar estratégias para utilização de transações e níveis de isolamento em SGBDs relacionais. As tarefas envolvem uma simulação de um sistema de reservas de passagem áreas.

Considere a seguinte tabela que registra os assentos reservados em um vôo:

**Assentos(num_voo,disp)**

onde **num_voo** um número inteiro de 1 a 200 e **disp** é um atributo booleano cujo valor é **true** se o assento estiver vago e **false** caso contrário. O valor inicial é **true**.

A reserva de um assento é feita em três passos:

>**Passo 1.**
O sistema recupera a lista dos assentos disponíveis.  
>**Passo 2.**
O cliente escolhe o assento. Esse passo deve ser simulado pela escolha aleatória de um dos assentos disponíveis, levando para isso um tempo de escolha de 1 segundo.  
>**Passo 3.**
O sistema registra a reserva do assento escolhido, atualizando o valor de **disp** para **false**.


Cada assento é reservado individualmente. Duas versões diferentes do processo de reserva devem ser implementadas.

>**Versão a.**
A reserva é implementada como uma única transação que inclui os três passos acima.  
>**Versão b.**
A reserva inclui uma transação para o Passo 1 e outra para o Passo 3. O Passo 2 não faz parte das transações, mas deve ser executado.


Agentes de viagens são responsáveis por realizar as reservas de 200 clientes no total. A atividade de um agente de viagens é simulada por uma thread.

Experimentos devem ser realizados simulando a atuação de k agentes de viagem trabalhando simultaneamente, onde k = 1,2,4,6,8 e 10. Cada agente/thread faz uma reserva de cada vez. As threads devem ser reiniciadas até que todos os 200 clientes tenham seus assentos reservados.

Dois conjuntos de experimentos devem ser feitos usando dois níveis de isolamento:  “read committed” e “serializable”.  Nos dois casos, o sistema deve ser configurado para realizar bloqueios a nível de tupla (linha).

As implementações devem ser feitas em Python3.8+ usando o SGBD PostgreSQL.

Considerando o descrito acima, execute as seguintes tarefas:


#### **Tarefa 17 -**

Implemente as versões a e b do processo de reserva.

**O que entregar:** Código fonte Python3.8+ documentado das duas versões.

#### **Tarefa 18 -**

Apresente gráficos de linha onde, para cada valor de k (número de agentes) no eixo x, temos no eixo y o tempo necessário para que todos os clientes efetuem suas reservas. Um gráfico diferente deve ser apresentado para cada par de versões da reserva e nível de isolamento.

#### **Tarefa 19 -**

Apresente uma tabela com o número máximo, mínimo e médio de vezes que um cliente teve que tentar reservar um assento até conseguir, ou seja, o número de vezes que uma reserva teve que ser refeita. A tabela considera as variações de k, versão de reserva e nível de isolamento.

#### **Tarefa 20 -**

Apresente uma análise dos resultados obtidos em cada versão de reserva e tipo de isolamento, explicando as diferenças entre resultados.